### DISCUSSION TITLE
Single-element lists should require a trailing comma (maybe)

### DISCUSSION THREAD

**Discussion by GrigorenkoPV** [hsutter/cppfront#1179] [Suggestions]
From what I can tell,
```cpp2
for 0..=10 do (i) {
	std::cout << i << std::endl;
}
```
iterates over a range of ints, 1 through 10, while
```cpp2
for (0..=10) do (i) {
	std::cout << i << std::endl;
}
```
iterates over a list of `cpp2::range<int>`s, which is probably due to the fact that `(0..=10)` gets treated as a single-element least.

I, personally, find it a bit confusing. I would not expect that adding a pair of parenthesis around an expression fundamentally changes it. And while I know that cpp2 currently uses `(...)` instead of `[...]` for lists, intuitively I do expect such behavior from `[...]`, but not from `(...)`.

For example, in Python, there are essentially two types of lists commonly used. The `[...]` "lists" (variable-length, a-la `std::vector`[^1]) and the `(...)` "tuples" (fixed length, a-la `std::array`[^1]). Python **allows** trailing commas (thankfully), but for single-element tuples it ***requires*** a trailing comma for disambiguation.
```python
def show(x):
    print(type(x).__name__, x)

show(  1     ) # int 1
show( (1)    ) # int 1
show( (1,)   ) # tuple (1,)
show( (1,2)  ) # tuple (1, 2)
show( (1,2,) ) # tuple (1, 2)
show( ()     ) # tuple ()
show( (,)    ) # SyntaxError: invalid syntax
show( [1]    ) # list [1]
show( [1,]   ) # list [1]
show( [1,2]  ) # list [1, 2]
show( [1,2,] ) # list [1, 2]
show( []     ) # list []
show( [,]    ) # SyntaxError: invalid syntax
```

So, should cpp2 do something similar?

[^1]: Except of course Python allows storing elements of different types, because it is dynamically typed. Essentially it stores a vector/array of `object`s.

**Comment by hsutter** [comment]
> From what I can tell,
> 
> ```
> for 0..=10 do (i) {
> 	std::cout << i << std::endl;
> }
> ```
> 
> iterates over a range of ints, 1 through 10, while

Yes.

> 
> ```
> for (0..=10) do (i) {
> 	std::cout << i << std::endl;
> }
> ```
> 
> iterates over a list of `cpp2::range<int>`s, which is probably due to the fact that `(0..=10)` gets treated as a single-element least.

Yes, the `(` `)` denote a list of values, and that's a list of one value which is a range.

You can find the same effect in non-range examples, such as `for r` vs `for (r)` where `r` is some container or range:

```
for  r  do (e) { } // ok, e is an element of r
for (r) do (e) { } // ok, e is r
```

Consider that we do want syntax like `(` `)` to make a list of things right there in the `for` statement:

```cpp
for (1, 2, 3) do (i) { } // ok
for (1, 2)    do (i) { } // ok
for (1)       do (i) { } // ok
for  1        do (i) { } // error, 1 is not a range
```

So the status quo for a `for` loop is that when you write a list of individual things you use `(` `)`, and when you write a range you use `...` or `..=`, and here's how they look side by side:

```cpp
for (1, 2, 3) do (i) { } // ok
for  1 ..= 3  do (i) { } // ok, same meaning
for  1 ... 4  do (i) { } // ok, same meaning
```

> Python **allows** trailing commas (thankfully), 

(BTW so does Cpp2 now, thanks again to everyone who motivated me to finally add that.)

> but for single-element tuples it _**requires**_ a trailing comma for disambiguation.

That's the kind of thing I would worry about.

**Comment by svew** [comment]
Remind again, was there a reason not to use `[ ]` to denote lists instead of `( )`, or just a choice made?

**Reply by sookach** [reply]
Just a choice made by Herb

**Comment by SebastianTroy** [comment]
It was a well reasoned choice though 😂 so yes, there were reasons, I was on the side of using [] initially, but I was converted, though it was a while ago so I don't remember specifics


On 2 August 2024 01:48:15 Andrew Sukach ***@***.***> wrote:

Just a choice made by Herb
